File: python_kata/src/data_structures/stack.py
from typing import Generic, TypeVar

T = TypeVar("T")

class SNode(Generic[T]):
    def __init__(self, val: T):
        self.val: T  = val
        self.prev: SNode[T] | None = None


class Stack(Generic[T]):
    head: SNode[T] | None
    length: int

    def __init__(self):
        self.head = None
        self.length = 0

    def peak(self) -> T | None:
        if not self.head:
            return None

        return self.head.val

    def push(self, item: T) -> None:
        node = SNode(item)
        if not self.head:
            self.head = node
            self.length += 1

            return None

        node.prev = self.head
        self.head = node

        self.length += 1

        return None

    def pop(self) -> T | None:
        if not self.head:
            return None

        node = self.head
        self.head = node.prev

        self.length -= 1

        # Free memory
        node.prev = None

        return node.val

File: typescript-kata/src/data-structures/stack.ts
class SNode<T> {
    val: T;
    prev: SNode<T> | undefined;

    constructor(val: T) {
        this.val = val;
        this.prev = undefined;
    }
}

export class Stack<T> {
    head: SNode<T> | undefined;
    length: number;

    constructor() {
        this.head = undefined;
        this.length = 0;
    }

    peak(): T | undefined {
        if (!this.head) {
            return undefined;
        }

        return this.head.val;
    }

    push(item: T): void {
        const node = new SNode(item);

        if (!this.head) {
            this.head = node;
            this.length++;

            return;
        }

        node.prev = this.head;
        this.head = node;
        this.length++

        return;
    }

    pop(): T | undefined {
        if (!this.head) {
            return undefined;
        }

        const node = this.head;
        this.head = node.prev;

        this.length--;

        // Free the memory
        node.prev = undefined;

        return node.val;
    }
}
